add custom ceph repo support

andrewschoen opened the pull request.

obnoxxx reviewed (commented):
Thank you for this PR, @andrewschoen !

This looks like a nice patchset but here are a few thoughts:

Firstly, this PR conceptually consists of two bigger changes: Adding a 4.21 specfile and adding support for IBM repos. Each of these changes alone looks big enough to me to justify splitting the change out into a separate PR. This is really a question and not a formal change request: how about splitting this PR into two PRs?

Secondly, I am wondering if the IBM repo support piece might be exposing too much information publicly. Please check and think about this. Happy to discuss. Again, not a change request (yet), but a question at this point.

andrewschoen commented:
> Thank you for this PR, @andrewschoen !
> 
> This looks like a nice patchset but here are a few thoughts:
> 
> Firstly, this PR conceptually consists of two bigger changes: Adding a 4.21 specfile and adding support for IBM repos. Each of these changes alone looks big enough to me to justify splitting the change out into a separate PR. This is really a question and not a formal change request: how about splitting this PR into two PRs?
> 

I would agree that these changes are conceptually two big and separate concerns. I needed them in one branch for my testing downstream though. I'd prefer to keep it one branch for now, if that becomes a blocker to merge I can split them though.

> Secondly, I am wondering if the IBM repo support piece might be exposing too much information publicly. Please check and think about this. Happy to discuss. Again, not a change request (yet), but a question at this point.

To me, this is less IBM support and more "custom repo" support. I could remove any mention of IBM here though and make this more generic. The gpg key is public information, I could remove that though and add another variable to control it. How does that sound?

obnoxxx commented:
@andrewschoen  wrote:
> > Thank you for this PR, @andrewschoen !
> > This looks like a nice patchset but here are a few thoughts:
> > Firstly, this PR conceptually consists of two bigger changes: Adding a 4.21 specfile and adding support for IBM repos. Each of these changes alone looks big enough to me to justify splitting the change out into a separate PR. This is really a question and not a formal change request: how about splitting this PR into two PRs?
> 
> I would agree that these changes are conceptually two big and separate concerns. I needed them in one branch for my testing downstream though. I'd prefer to keep it one branch for now, if that becomes a blocker to merge I can split them though.

Not a blocker for me. Maybe consider more piecemeal PRs for future contributions?

> 
> > Secondly, I am wondering if the IBM repo support piece might be exposing too much information publicly. Please check and think about this. Happy to discuss. Again, not a change request (yet), but a question at this point.
> 
> To me, this is less IBM support and more "custom repo" support. I could remove any mention of IBM here though and make this more generic. The gpg key is public information, I could remove that though and add another variable to control it. How does that sound?

Sounds good to me to make the change more generic this way.

anoopcs9 commented:
I like the idea of introducing an optional support to use a custom base url for ceph packages. May be we could generalize to consume such a custom url in our default ceph dev repo templates? Right now `ceph_base_url` is tied to just ibm ceph repo template.

andrewschoen commented:
> @andrewschoen wrote:
> 
> > > Thank you for this PR, @andrewschoen !
> > > This looks like a nice patchset but here are a few thoughts:
> > > Firstly, this PR conceptually consists of two bigger changes: Adding a 4.21 specfile and adding support for IBM repos. Each of these changes alone looks big enough to me to justify splitting the change out into a separate PR. This is really a question and not a formal change request: how about splitting this PR into two PRs?
> > 
> > 
> > I would agree that these changes are conceptually two big and separate concerns. I needed them in one branch for my testing downstream though. I'd prefer to keep it one branch for now, if that becomes a blocker to merge I can split them though.
> 
> Not a blocker for me. Maybe consider more piecemeal PRs for future contributions?

I can do that, thanks for your understanding in this case.

andrewschoen commented:
> I like the idea of introducing an optional support to use a custom base url for ceph packages. May be we could generalize to consume such a custom url in our default ceph dev repo templates? Right now `ceph_base_url` is tied to just ibm ceph repo template.

@anoopcs9 I've added a new commit for a more general custom ceph repo support. Are you suggesting that instead of a spearate templates move this logic into the existing dev repo templates? I think that could work.

anoopcs9 commented:
> > I like the idea of introducing an optional support to use a custom base url for ceph packages. May be we could generalize to consume such a custom url in our default ceph dev repo templates? Right now `ceph_base_url` is tied to just ibm ceph repo template.
> 
> @anoopcs9 I've added a new commit for a more general custom ceph repo support. Are you suggesting that instead of spearate templates move this logic into the existing dev repo templates? I think that could work.

Yeah, and may be remove the `dev` keyword to better the suit the intent? Thus we end up using the custom url(and probably gpg key) whenever it is defined or the default dev repo url otherwise, all based out of same repo template.

andrewschoen commented:
> > > I like the idea of introducing an optional support to use a custom base url for ceph packages. May be we could generalize to consume such a custom url in our default ceph dev repo templates? Right now `ceph_base_url` is tied to just ibm ceph repo template.
> > 
> > 
> > @anoopcs9 I've added a new commit for a more general custom ceph repo support. Are you suggesting that instead of spearate templates move this logic into the existing dev repo templates? I think that could work.
> 
> Yeah, and may be remove the `dev` keyword to better the suit the intent? Thus we end up using the custom url(and probably gpg key) whenever it is defined or the default dev repo url otherwise, all based out of same repo template.

I agree, I'll make that change.

anoopcs9 commented:
> > Thank you for this PR, @andrewschoen !
> > This looks like a nice patchset but here are a few thoughts:
> > Firstly, this PR conceptually consists of two bigger changes: Adding a 4.21 specfile and adding support for IBM repos. Each of these changes alone looks big enough to me to justify splitting the change out into a separate PR. This is really a question and not a formal change request: how about splitting this PR into two PRs?
> 
> I would agree that these changes are conceptually two big and separate concerns. I needed them in one branch for my testing downstream though. I'd prefer to keep it one branch for now, if that becomes a blocker to merge I can split them though.

Samba 4.21 is already out and we would like to get the corresponding spec file added for CI Jobs to pick it up. Let me know if I should split the relevant change(obviously retaining the ownership) to get it done.

andrewschoen commented:
> > > Thank you for this PR, @andrewschoen !
> > > This looks like a nice patchset but here are a few thoughts:
> > > Firstly, this PR conceptually consists of two bigger changes: Adding a 4.21 specfile and adding support for IBM repos. Each of these changes alone looks big enough to me to justify splitting the change out into a separate PR. This is really a question and not a formal change request: how about splitting this PR into two PRs?
> > 
> > 
> > I would agree that these changes are conceptually two big and separate concerns. I needed them in one branch for my testing downstream though. I'd prefer to keep it one branch for now, if that becomes a blocker to merge I can split them though.
> 
> Samba 4.21 is already out and we would like to get the corresponding spec file added for CI Jobs to pick it up. Let me know if I should split the relevant change(obviously retaining the ownership) to get it done.

@anoopcs9 I've removed my 4.21 spec commit from this PR. Please feel free to add it back in a separate PR.

I've been out sick, I do plan to address your PR comments here soon. Thanks!

anoopcs9 commented:
> > > > Thank you for this PR, @andrewschoen !
> > > > This looks like a nice patchset but here are a few thoughts:
> > > > Firstly, this PR conceptually consists of two bigger changes: Adding a 4.21 specfile and adding support for IBM repos. Each of these changes alone looks big enough to me to justify splitting the change out into a separate PR. This is really a question and not a formal change request: how about splitting this PR into two PRs?
> > > 
> > > 
> > > I would agree that these changes are conceptually two big and separate concerns. I needed them in one branch for my testing downstream though. I'd prefer to keep it one branch for now, if that becomes a blocker to merge I can split them though.
> > 
> > 
> > Samba 4.21 is already out and we would like to get the corresponding spec file added for CI Jobs to pick it up. Let me know if I should split the relevant change(obviously retaining the ownership) to get it done.
> 
> @anoopcs9 I've removed my 4.21 spec commit from this PR. Please feel free to add it back in a separate PR.

Cool. #60 has been created to add the 4.21 spec file.

> I've been out sick, I do plan to address your PR comments here soon. Thanks!

Great, thanks.

obnoxxx reviewed (changes requested):
Thanks for the update, @andrewschoen .

The patchset looks nice, but I have small requests:

1. please  add reasonable default values to the new variables?
2. please explain the purpose of the new vars in the README and give examples.

anoopcs9 commented on ansible/roles/prep.dirs/files/ceph-el9-repo.tpl.j2:
```diff
@@ -0,0 +1,13 @@
+config_opts['yum.conf'] += """
+
+[ceph-custom-repo]
+name = Ceph 
```
```suggestion
[ceph-repo]
name = Ceph builds
```

anoopcs9 commented on ansible/roles/prep.dirs/tasks/repo.yml:
```diff
@@ -9,25 +9,31 @@
     loop_var: mock_config
 
-- name: adapt mock config to use additional template for ceph
+- name: adapt mock config to use additional dev template for ceph
```
We may not need this change anymore.

anoopcs9 commented on ansible/roles/prep.dirs/tasks/repo.yml:
```diff
@@ -9,25 +9,31 @@
     - name: determine ceph repo base url
       shell: curl -s "https://shaman.ceph.com/api/search/?project=ceph&distros=centos/{{ item }}/x86_64&flavor=default&ref=main&sha1=latest" | jq -r .[0].url
       register: get_ceph_repo_url
+      when: ceph_base_url is not defined
 
     - set_fact:
         "ceph_el{{ item }}_repo_base_url": "{{ get_ceph_repo_url.stdout }}"
+      when: ceph_base_url is not defined
```
You could go for a nested block here to group both tasks when `ceph_base_url` is not defined.

andrewschoen replied:
Good idea, I've made that change.

anoopcs9 reviewed (changes requested):
I think [ceph-el9-custom-repo.tpl.j2](https://github.com/samba-in-kubernetes/samba-build/pull/57/files#diff-c6f32878a9e43db9c236cbfb7aa16d3ea9ecc22e9b22157649139d0882b23737) is a left over, you can remove it.

obnoxxx commented:
@andrewschoen  wrote:
> > Thanks for the update, @andrewschoen .
> > The patchset looks nice, but I have small requests:
> > 
> > 1. please  add reasonable default values to the new variables?
> 
> I think in this case the most reasonable default might be undefined. I could default to `none` and change the logic in the templates accordingly. How does that sound?

even the term "custom ceph repo" seems to suggest to me  that there should be a non-none default for the ceph repo (probably an upstream one). this is why I am asking for a  default for the new setting. I don't really mind whether this is achieved  by a variable's default valu or  by  template logic. 


> > 2. please explain the purpose of the new vars in the README and give examples.
> 
> Of course, I can add those docs.

sounds good. looking forward to reading them.

andrewschoen commented:
> I think [ceph-el9-custom-repo.tpl.j2](https://github.com/samba-in-kubernetes/samba-build/pull/57/files#diff-c6f32878a9e43db9c236cbfb7aa16d3ea9ecc22e9b22157649139d0882b23737) is a left over, you can remove it.

@anoopcs9  Good catch, I've removed it. Thanks for the review, I think I've got all your suggestions made in my latest commits.

andrewschoen commented:
> @andrewschoen wrote:
> 
> > > Thanks for the update, @andrewschoen .
> > > The patchset looks nice, but I have small requests:
> > > 
> > > 1. please  add reasonable default values to the new variables?
> > 
> > 
> > I think in this case the most reasonable default might be undefined. I could default to `none` and change the logic in the templates accordingly. How does that sound?
> 
> even the term "custom ceph repo" seems to suggest to me that there should be a non-none default for the ceph repo (probably an upstream one). this is why I am asking for a default for the new setting. I don't really mind whether this is achieved by a variable's default valu or by template logic.
> 

I added a default value of `false` for both `ceph_gpgkey` and `ceph_base_url`. This allows us to keep the original default action here which is to use the latest ceph dev build from shaman. I didn't want to add an upstream repo url here as I thought it might complicate the logic in the tasks/templates. How does that sound?

> > > 2. please explain the purpose of the new vars in the README and give examples.
> > 
> > 
> > Of course, I can add those docs.
> 
> sounds good. looking forward to reading them.

I've added docs for `ceph_repo_url` and `ceph_gpgkey` usage. I've also added a small section explaning a bit about how this project uses Ceph. @obnoxxx could you please review those?

obnoxxx commented on ansible/roles/prep.dirs/files/ceph-el8-repo.tpl.j2:
```diff
@@ -0,0 +1,13 @@
+
+[ceph-repo]
+name = Ceph builds
+baseurl={{ ceph_el8_repo_base_url }}/x86_64
```
should this be `ceph_base_url`?

andrewschoen replied:
This is right, `ceph_el8_repo_base_url` gets set to `ceph_base_url` in prep.dirs/tasks/repo.yml if it's provided, or the appropriate url from shaman otherwise. `ceph_base_url` acts as an override for all versions.

obnoxxx reviewed (commented):
I found a use of `ceph_el8_repo_base_url`which might need to be `ceph_base_url`.

Also wondering if  `ceph_repo_base_url` might be a better var name than `ceph_base_url`

andrewschoen commented:
> Also wondering if `ceph_repo_base_url` might be a better var name than `ceph_base_url`

I don't have a strong opinion on this, I wouldn't mind changing it to `ceph_repo_base_url` to be make it more clear if that makes more sense.

I've been wondering if we can get rid of the idea of version specific templates here entirely though. In practice, we only build for one version at a time so why generate configs/repos for multiple versions in a single run? Everything in the repo templates are the same right now besides the baseurl. @anoopcs9 any thoughts on this?

anoopcs9 commented:
/test all

anoopcs9 commented on ansible/roles/prep.dirs/tasks/repo.yml:
```diff
@@ -4,38 +4,40 @@
   with_items:
-    - "{{ mock_dir }}/centos-stream+epel-next-{{ item }}-x86_64.cfg"
-    - "{{ mock_dir }}/rhel-{{ item }}-x86_64.cfg"
+    - "{{ mock_dir }}/centos-stream+epel-next-{{ os_vers }}-x86_64.cfg"
```
This would mean that we search for _centos-stream+epel-next-40-x86_64.cfg_ when invoked with `make rpms.fedora`(or _centos-stream+epel-next-39-x86_64.cfg_ when invoked with `make rpms.fedora vers=39`), right? Similarly for _rhel_ templates.

andrewschoen replied:
Yeah, that's right. We use the value of os_vers, which means that we only generate configs for the os version we're building for. This value is set in the playbooks. Do we generate fedora and rhel config when building for fedora?

andrewschoen replied:
Ok, I do see what you mean about how this would look for a file that doesn't exist when building for fedora. Honestly, the entirely of repo.yml isn't necessary for fedora though. Maybe we can skip this if we're not building for centos or rhel?

This also fails if ran from a playbook that does not set `os_vers`, like ansible/build.srpm.yml. I'd have to figure out a way around that as well.

anoopcs9 commented on ansible/roles/prep.dirs/tasks/main.yml:
```diff
@@ -50,6 +49,3 @@
     - "/etc/mock/fedora-*-x86_64.cfg"
 
 - include_tasks: repo.yml
-  with_items:
-    - "8"
-    - "9"
```
I think we should leave it untouched. See below.

andrewschoen replied:
Removing this allows us to only create configs for the os version being built for. If we leave this, we get centos8 configs generated when we are building centos9 rpms, I was trying to avoid this and only build for one specfic os version at a time.

anoopcs9 replied:
Alright, now it makes sense.

anoopcs9 commented on ansible/roles/prep.dirs/tasks/repo.yml:
```diff
@@ -4,38 +4,40 @@
       template:
-        src: "{{ mock_dir }}/ceph-el{{ item }}-dev-repo.tpl.j2"
-        dest: "{{ mock_dir }}/ceph-el{{ item }}-dev-repo.tpl"
+        src: "{{ mock_dir }}/ceph-el{{ os_vers }}-repo.tpl.j2"
```
Did you mean _ceph-repo.tpl.j2_?

But then I see a problem. We will then end up with the same base url for every os version template. Let's think on a way out.

anoopcs9 commented on ansible/vars.yml:
```diff
@@ -5,3 +5,5 @@ srpm_dir: "{{ build_dir }}/srpms"
 rpms_dir: "{{ build_dir }}/rpms"
 mock_dir: "{{ build_dir }}/mock"
 git_repo_url: "https://git.samba.org/samba.git"
+ceph_base_url: false
+ceph_gpgkey: false
```
Let's not make it a boolean. We expect a string, defined or not.

obnoxxx replied:
@anoopcs9  wrote:
> Let's not make it a boolean.

I agree, and this confirms an earlier comment of mine.

>We expect a string, defined or not.

do you mean, ëmpty or not"? because empty string is defined,
 but could be considered undefined or not specified.

anoopcs9 replied:
Correct.

anoopcs9 commented on ansible/roles/prep.dirs/tasks/repo.yml:
```diff
@@ -4,38 +4,40 @@
-        "ceph_el{{ item }}_repo_base_url": "{{ get_ceph_repo_url.stdout }}"
+      - set_fact:
+          "ceph_base_url": "{{ get_ceph_repo_url.stdout }}"
+      when: not ceph_base_url
```
I would recommend `when: ceph_base_url is not defined` from your previous iteration. See related comment towards the end.

anoopcs9 reviewed (changes requested):
> > Also wondering if `ceph_repo_base_url` might be a better var name than `ceph_base_url`

> I don't have a strong opinion on this, I wouldn't mind changing it to `ceph_repo_base_url` to be make it more clear if that makes more sense.

`ceph_repo_base_url` seems to be more complete. May be we rename `ceph_gpgkey` accordingly?

> I've been wondering if we can get rid of the idea of version specific templates here entirely though. In practice, we only build for one version at a time so why generate configs/repos for multiple versions in a single run? Everything in the repo templates are the same right now besides the baseurl. @anoopcs9 any thoughts on this?

That would be a nice improvement. I think I've made some comments around this area below.

anoopcs9 commented on ansible/roles/prep.dirs/tasks/repo.yml:
```diff
@@ -4,38 +4,40 @@
-        "ceph_el{{ item }}_repo_base_url": "{{ get_ceph_repo_url.stdout }}"
+      - set_fact:
+          "ceph_repo_base_url": "{{ get_ceph_repo_url.stdout }}"
+      when: not ceph_repo_base_url
```
```suggestion
      when: ceph_repo_base_url is not defined
```

obnoxxx commented on ansible/vars.yml:
```diff
@@ -5,3 +5,5 @@ srpm_dir: "{{ build_dir }}/srpms"
 rpms_dir: "{{ build_dir }}/rpms"
 mock_dir: "{{ build_dir }}/mock"
 git_repo_url: "https://git.samba.org/samba.git"
+ceph_repo_base_url: false
+ceph_repo_gpgkey: false
```
@anoopcs9  I do agree to this, but we could also fix this as a follow-up change.

obnoxxx replied:
But preferable to have it from first merging on.

obnoxxx reviewed (changes requested):
This is almost good to go in now.

Please address @anoopcs9's outstanding requests.

If it is not Asking for too much, I'd prefer the variables to be introduced withe final name in the introducing commit instead of having a renaming commit at the end of the patchset. IMO this would result in a more natural commit flow

andrewschoen commented:
> This is almost good to go in now.
> 
> Please address @anoopcs9's outstanding requests.
> 
> If it is not Asking for too much, I'd prefer the variables to be introduced withe final name in the introducing commit instead of having a renaming commit at the end of the patchset. IMO this would result in a more natural commit flow

@obnoxxx @anoopcs9 thank you both for your reviews. I think I've got all feedback incorporated and the commit history cleaned up now.

I had to add a default for `os_vers` in prep.dirs/tasks/repo.yml so that playbooks that don't set `os_vers` can still run. Would you rather that be set at the playbook level for all playbooks instead?

I've also added a tag of `repos` to the include for prep.dirs/tasks/repo.yml so that the fedora focused playbooks can skip these tasks. Otherwise they fail trying to inlude a config file that does not exist because of the value for `os_vers` set by those playbooks.

anoopcs9 reviewed (changes requested):
> I had to add a default for os_vers in prep.dirs/tasks/repo.yml so that playbooks that don't set os_vers can still run. Would you rather that be set at the playbook level for all playbooks instead?

I think most of the playbooks already set the default. You can update the missing ones at the playbook level.

Otherwise lgtm.

andrewschoen commented:
> > I had to add a default for os_vers in prep.dirs/tasks/repo.yml so that playbooks that don't set os_vers can still run. Would you rather that be set at the playbook level for all playbooks instead?
> 
> I think most of the playbooks already set the default. You can update the missing ones at the playbook level.

Sounds good, I've made this change.

anoopcs9 reviewed (approved):
Thanks for your patience, which is really appreciated. It looks much cleaner now.

obnoxxx reviewed (approved):
Thanks for the clean-up and for addressing the requests.

 LGTM now !
